.github/workflows/test-tpg.yml

```diff
@@ -14,36 +20,89 @@ on:
         required: true
 
+concurrency:
+  group: test-tpg-${{ github.event.inputs.owner }}-${{ github.event.inputs.repo }}-${{ github.event.inputs.branch }}
+  cancel-in-progress: true
+
 jobs:
-  build-and-test:
+  build-and-unit-test:
     runs-on: ubuntu-latest
+    timeout-minutes: 30
     steps:
     - name: Checkout Repository
       uses: actions/checkout@v2
       with:
-        repository: ${{ github.event.client_payload.owner }}/${{ github.event.client_payload.repo }}
-        ref: ${{ github.event.client_payload.branch }}
+        repository: ${{ github.event.inputs.owner }}/${{ github.event.inputs.repo }}
+        ref: ${{ github.event.inputs.branch }}
         path: provider
+        fetch-depth: 2
+    - name: Check for Code Changes
+      id: pull_request
+      run: |
+        cd provider
+        gofiles=$(git diff --name-only HEAD~1 | { grep -e "\.go$" -e "go.mod$" -e "go.sum$" || test $? = 1; })
+        if [ -z "$gofiles" ]; then
+          echo "has_changes=false" >> $GITHUB_OUTPUT
+        else
+          echo "has_changes=true" >> $GITHUB_OUTPUT
+        fi
+    - name: Get Job URL
+      if: ${{ !failure() && steps.pull_request.outputs.has_changes == 'true' }}
+      id: get_job
+      run: |
+        response=$(curl --get -Ss -H "Authorization: token ${{ secrets.GITHUB_TOKEN }}" -H "Accept: application/vnd.github.v3+json" "https://api.github.com/repos/${{ github.repository }}/actions/runs/${{ github.run_id }}/attempts/${{ github.run_attempt }}/jobs")
+        html_url=$(echo "$response" | jq -r --arg job_name "${{ github.job }}" '.jobs | map(select(.name == $job_name)) | .[0].html_url')
+        echo "url=${html_url}" >> $GITHUB_OUTPUT
+    - name: Post Pending Status to Pull Request
+      if: ${{ !failure() && steps.pull_request.outputs.has_changes == 'true' }}
+      run: |
+        curl -X POST -H "Authorization: token ${{secrets.GITHUB_TOKEN}}" \
+        -H "Accept: application/vnd.github.v3+json" \
+        "https://api.github.com/repos/GoogleCloudPlatform/magic-modules/statuses/${{github.event.inputs.sha}}" \
+        -d '{
+        "context": "${{ github.event.inputs.repo }}${{ env.status_suffix }}",
+        "target_url": "${{ steps.get_job.outputs.url }}",
+        "state": "pending"
+        }'
     - name: Set up Go
-      uses: golang/setup-go@v2
+      if: ${{ !failure() && steps.pull_request.outputs.has_changes == 'true' }}
+      uses: actions/setup-go@v4
       with:
-        go-version: '1.19.x'
-    - name: Build
+        go-version: '^1.19.1'
+    - name: Build Provider
+      if: ${{ !failure() && steps.pull_request.outputs.has_changes == 'true' }}
       run: |
         cd provider
         go build
-    - name: Lint
+    - name: Run Unit Tests
+      if: ${{ !failure() && steps.pull_request.outputs.has_changes == 'true' }}
+      run: |
+        cd provider
+        make test
+    - name: Lint Check
+      if: ${{ !cancelled() && steps.pull_request.outputs.has_changes == 'true' }}
```

Comment:
I think `cancelled()` probably does the right thing, but possibly `success() || failure()` might be more clear - it seems like that might more accurately reflect what you're going for & is recommended in the GitHub docs: https://docs.github.com/en/actions/learn-github-actions/expressions#always What would you think about that?

Reply:
I believe this may be correct. I'll have to do research

Reply:
I've thought a lot about this. I think my approach is correct.

presumably the statuses we can check for are `success` `failure` and `canceled`. `!canceled()` logically is the most condensed and direct form. 

I'm not sure if `success || failure` will return `false || false` in case of cancelation. I would assume so but `!canceled` seems direct to the usecase I want to monitor for. 

I commented on some external discussion but disagreed with their conclusions.
https://github.com/orgs/community/discussions/26303#discussioncomment-5821344

resolving this for now